pkg/raft/storage.go

```diff
@@ -73,25 +66,41 @@ type Storage interface {
 	// encountered an unavailable entry in [lo, hi).
 	Entries(lo, hi, maxSize uint64) ([]pb.Entry, error)
 
-	// Term returns the term of entry i, which must be in the range
-	// [FirstIndex()-1, LastIndex()]. The term of the entry before
-	// FirstIndex is retained for matching purposes even though the
-	// rest of that entry may not be available.
+	// Term returns the term of the entry at the given index, which must be in the
+	// valid range: [FirstIndex()-1, LastIndex()]. The term of the entry before
+	// FirstIndex is retained for matching purposes even though the rest of that
+	// entry may not be available.
 	Term(i uint64) (uint64, error)
 	// LastIndex returns the index of the last entry in the log.
 	LastIndex() (uint64, error)
-	// FirstIndex returns the index of the first log entry that is
-	// possibly available via Entries (older entries have been incorporated
-	// into the latest Snapshot; if storage only contains the dummy entry the
-	// first log entry is not available).
+	// FirstIndex returns the index of the first log entry that is possibly
+	// available via Entries (older entries have been incorporated into the
+	// StateStorage.Snapshot; if storage only contains the dummy entry or initial
+	// snapshot then the first log entry is not available).
```

Comment:
I think it's 101, will check and add a comment.

But in general, I think a better API would be parallel to `logSlice` (any log storage is just a `logSlice` with pagination):

- `Prev()` returns `entryID` of the entry which starts the log. This entry is not stored, but we know its ID.
- `LastIndex` / `Entries` / `Term[At]` would be the same.

This, together with the `(lo, hi]` semantics instead of `[lo, hi)` would eliminate the last cases with weird +-1 arithmetics.

Reply:
Done.